tachyon/crypto/hashes/sponge/poseidon2/icicle/icicle_poseidon2.h

```diff
@@ -0,0 +1,138 @@
+#ifndef TACHYON_CRYPTO_HASHES_SPONGE_POSEIDON2_ICICLE_ICICLE_POSEIDON2_H_
+#define TACHYON_CRYPTO_HASHES_SPONGE_POSEIDON2_ICICLE_ICICLE_POSEIDON2_H_
+
+#include <memory>
+
+#include "absl/types/span.h"
+#include "third_party/icicle/include/hash/hash_config.h"
+
+#include "tachyon/base/logging.h"
+#include "tachyon/crypto/hashes/sponge/poseidon2/poseidon2_vendor.h"
+#include "tachyon/device/gpu/gpu_device_functions.h"
+#include "tachyon/export.h"
+#include "tachyon/math/elliptic_curves/bn/bn254/fr.h"
+#include "tachyon/math/finite_fields/baby_bear/baby_bear.h"
+
+namespace tachyon::crypto {
+
+template <class F>
+struct IsIciclePoseidon2SupportedImpl {
+  constexpr static bool value = false;
+};
+
+template <>
+struct IsIciclePoseidon2SupportedImpl<math::BabyBear> {
+  constexpr static bool value = true;
+};
+
+template <>
+struct IsIciclePoseidon2SupportedImpl<math::bn254::Fr> {
+  constexpr static bool value = true;
+};
+
+template <typename F>
+constexpr bool IsIciclePoseidon2Supported =
+    IsIciclePoseidon2SupportedImpl<F>::value;
+
+struct TACHYON_EXPORT IciclePoseidon2Options {
+  bool are_inputs_on_device = false;
+  bool are_outputs_on_device = false;
+  bool is_async = false;
+};
+
+template <typename F>
+class IciclePoseidon2 {
+ public:
+  IciclePoseidon2(
+      gpuMemPool_t mem_pool, gpuStream_t stream,
+      const IciclePoseidon2Options& options = IciclePoseidon2Options())
+      : mem_pool_(mem_pool), stream_(stream) {
+    ::device_context::DeviceContext ctx{stream_, /*device_id=*/0, mem_pool_};
+    config_.reset(new ::hash::HashConfig{
+        ctx,
+        options.are_inputs_on_device,
+        options.are_outputs_on_device,
+        options.is_async,
+    });
+    VLOG(1) << "IciclePoseidon2 is created";
+  }
+  IciclePoseidon2(const IciclePoseidon2& other) = delete;
+  IciclePoseidon2& operator=(const IciclePoseidon2& other) = delete;
+  ~IciclePoseidon2() { CHECK(Delete()); }
+
+  void* impl() { return impl_; }
+  const void* impl() const { return impl_; }
+
+  [[nodiscard]] bool Create(unsigned int rate, unsigned int width,
```

Comment:
Can we not use `Poseidon2Params` here somehow?

Reply:
We can't use `Poseidon2Params` because, to improve build performance and reduce compile times for CUDA code, we rely on template specializations. This method doesn't allow for the use of additional template parameters.